codec-http2/src/main/java/io/netty/handler/codec/http2/CleartextHttp2ServerUpgradeHandler.java

```diff
@@ -0,0 +1,100 @@
+ * This handler config pipeline for h2c upgrade when handler added.
+ * And will update pipeline once it detect the connection is starting HTTP/2 by
+ * prior knowledge or not.
+ */
+@UnstableApi
+public final class CleartextHttp2ServerUpgradeHandler extends ChannelHandlerAdapter {
+    private static final ByteBuf CONNECTION_PREFACE = unreleasableBuffer(connectionPrefaceBuf());
+
+    private final HttpServerCodec httpServerCodec;
+    private final HttpServerUpgradeHandler httpServerUpgradeHandler;
+    private final ChannelHandler http2ServerHandler;
+
+    /**
+     * Creates the channel handler provide cleartext HTTP/2 upgrade from HTTP
+     * upgrade or prior knowledge
+     *
+     * @param httpServerCodec the http server codec
+     * @param httpServerUpgradeHandler the http server upgrade handler for HTTP/2
+     * @param http2ServerHandler the http2 server handler, will be added into pipeline
+     *                           when starting HTTP/2 by prior knowledge
+     */
+    public CleartextHttp2ServerUpgradeHandler(HttpServerCodec httpServerCodec,
+                                              HttpServerUpgradeHandler httpServerUpgradeHandler,
+                                              ChannelHandler http2ServerHandler) {
+        this.httpServerCodec = httpServerCodec;
+        this.httpServerUpgradeHandler = httpServerUpgradeHandler;
+        this.http2ServerHandler = http2ServerHandler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline()
+           .addBefore(ctx.name(), null, new PriorKnowlegeHandler())
+           .addBefore(ctx.name(), null, httpServerCodec)
+           .addBefore(ctx.name(), null, httpServerUpgradeHandler)
+           .remove(this);
+    }
+
+    /**
+     * Peek inbound message to determine current connection wants to start HTTP/2
+     * by HTTP upgrade or prior knowledge
+     */
+    private final class PriorKnowlegeHandler extends ByteToMessageDecoder {
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+            int prefaceLength = CONNECTION_PREFACE.readableBytes();
+            int bytesRead = Math.min(in.readableBytes(), prefaceLength);
+
+            if (!ByteBufUtil.equals(CONNECTION_PREFACE, CONNECTION_PREFACE.readerIndex(),
+                                    in, in.readerIndex(), bytesRead)) {
+                ctx.pipeline().remove(this);
+            } else if (bytesRead == prefaceLength) {
+                // Full h2 preface match, removed source codec, using http2 codec to handle
+                // following network traffic
+                ctx.pipeline().remove(httpServerCodec);
+                ctx.pipeline().remove(httpServerUpgradeHandler);
+
+                ctx.pipeline().addAfter(ctx.name(), null, http2ServerHandler);
+
+                ctx.pipeline().remove(this);
+            } else {
+                // Wait for further bytes
```

Comment:
Done by https://github.com/netty/netty/pull/6496/commits/20bee40963957cb4a4bd1aa114699a48bc53e0c4 (I forgot this one as other reviewers had point out this problem...)